File: algorithms/1331.rank-transform-of-an-array/solution.rb
# @param {Integer[]} arr
# @return {Integer[]}
def array_rank_transform(arr)
  rank = {}
  arr.sort.uniq.each_with_index do |score, idx|
    rank[score] = idx + 1
  end
  arr.map {|score| rank[score]}
end

File: algorithms/1331.rank-transform-of-an-array/solution.rs
impl Solution {
    pub fn array_rank_transform(arr: Vec<i32>) -> Vec<i32> {
        use std::collections::HashMap;

        let mut sorted_scores: Vec<i32> = arr.clone();
        sorted_scores.sort_unstable();
        sorted_scores.dedup();

        let mut rank: HashMap<i32, i32> = sorted_scores
            .iter()
            .enumerate()
            .map(|(idx, &score)| {
                (score, idx as i32 + 1)
            })
            .collect();

        arr
            .iter()
            .map(|score| {
                rank[score]
            })
            .collect()
    }
}
